dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/action/validation/BaseProgramRuleActionValidator.java

```diff
@@ -110,15 +114,7 @@ private ProgramRuleActionValidationResult handleDataElement(
           .build();
     }
 
-    List<ProgramStage> stages = validationContext.getProgramStages();
-
-    if (stages == null || stages.isEmpty()) {
-      stages =
-          validationContext
-              .getProgramRuleActionValidationService()
-              .getProgramStageService()
-              .getProgramStagesByProgram(program);
-    }
+    List<ProgramStage> stages = getRequiredProgramStages(validationContext, program);
```

Comment:
I would add a comment here explaining why you are loading the missing program stages from the DB even though they are present in the program, as you explained in slack.

Reply:
fixed